lib/internal/test_runner/tap_lexer.js

```diff
@@ -0,0 +1,589 @@
+  }
+
+  next() {
+    const char = this.peek();
+    if (char === undefined) {
+      return undefined;
+    }
+
+    this.pos++;
+    this.column++;
+    if (char === '\n') {
+      this.line++;
+      this.column = 0;
+    }
+
+    return char;
+  }
+
+  error(message, token, expected) {
+    const COLOR_UNDERLINE = '\u001b[4m';
+    const COLOR_WHITE = '\u001b[30;1m';
+    const COLOR_RED = '\u001b[31m';
+    const COLOR_RESET = '\u001b[0m';
+    const COLOR_YELLOW_BG = '\u001b[43m';
+
+    const { column, line } = token.location;
+    const indent = column + ('' + line).length + 1;
+
+    const sourceLines = this.input.split('\n');
+    const sourceLine = sourceLines[line - 1];
+
+    // highlight in red the token
+    const sourceLineWithToken =
+      sourceLine.slice(0, column - 1) +
+      COLOR_UNDERLINE +
+      COLOR_RED +
+      sourceLine.slice(column - 1, column + (token.value.length - 1)) +
+      COLOR_RESET +
+      sourceLine.slice(column + (token.value.length - 1));
+
+    console.error(
+      `${COLOR_YELLOW_BG + COLOR_RED}TAP Syntax Error:${COLOR_RESET}`
+    );
+
+    sourceLines.forEach((sourceLine, index) => {
+      const leadingZero = index < 9 ? '0' : '';
+      if (index === line - 1) {
+        console.error(
+          `${COLOR_RED}${leadingZero}${index + 1}: ${sourceLine + COLOR_RESET}`
+        );
+      } else {
+        console.error(
+          `${COLOR_WHITE}${leadingZero}${
+            index + 1
+          }:${COLOR_RESET} ${sourceLine}`
+        );
+      }
+    });
+
+    console.error(`\n${COLOR_YELLOW_BG + COLOR_RED}Diagnostic:${COLOR_RESET}`);
+
+    console.error(`${COLOR_WHITE}${line}:${COLOR_RESET} ${sourceLineWithToken}
+${COLOR_RED}${' '.repeat(indent)}│${COLOR_RESET}
+${COLOR_RED}${' '.repeat(indent)}└┤${message}, got "${
+      token.value
+    }" at line ${line}, column ${column} (start ${token.location.start}, end ${
+      token.location.end + 1
+    })${COLOR_RESET}
+      `);
+    process.exit(1);
+  }
+}
+
+class TapLexer {
+  static Keywords = new Set([
+    'TAP',
+    'version',
+    'ok',
+    'not',
+    '...',
+    '---',
+    '..',
+    'pragma',
+    '-',
+    '+',
+
+    // NOTE: "Skip", "Todo" and "Bail out!" literals are deferred to the parser
+  ]);
+
+  isComment = false;
+
+  constructor(source) {
+    this.source = new InputStream(source);
+    this.index = 0;
+    this.line = 1;
+    this.column = 0;
+
+    this.escapeStack = [];
+
+    this.lastScannedToken = new Token({
+      kind: TokenKind.EOL,
+      value: TokenKind.EOL,
+      stream: this.source,
+    });
+  }
+
+  *scan() {
+    while (!this.source.eof()) {
+      let token = this.scanToken();
+
+      // remember the last scanned token (except for whitespace)
+      if (token.kind !== TokenKind.WHITESPACE) {
+        this.lastScannedToken = token;
+      }
+      yield token;
+    }
+
+    yield this.scanEOF();
+  }
+
+  *scanAll() {
+    for (const token of this.scan()) {
+      yield token;
+    }
```

Comment:
Why does it need to be a separated method? Couldn't we reuse the same method?
```suggestion
    yield* this.scan();
```

Reply:
Thanks for pointing this. There is actually no particular reason. I had this method initially while doing some tests and I forget to remote it. I might remove it and just use scan() instead.